### PR TITLE
fix #8389 - keep user logged in

### LINKED ISSUE #8389: Keep the user logged in
Our checkout modal currently [does not keep the user](https://github.com/unlock-protocol/unlock/blob/master/unlock-app/src/pages/checkout.js#L19) "logged in" (i.e. it does not remember the wallet used to connect and/or the Unlock account between sessions). We need to change this. 

The reason we enabled `skipAutoLogin` is because it was creating issues when the checkout modal was configured to request the user to sign. Basically, it was skipping the request to sign even though it is required by the configuration because the signature request is too tightly coupled with the logic to connect the wallet.

So a first step should be to "isolate" the requirement to sign from the "connect" wallet. This way we can prompt the user to sign something even when they are already logged in.

A second issue will be for Unlock account users. When these users log-in we load the private key from the server and use the password to decrypt it on the front-end. The private key is then in memory for the duration of the session but if the session is terminated (tab closed), then the private key is "flushed". This means we currently cannot keep the Unlock users "logged in", even though that would be a useful step. We need to identify a way to do that.

### PR DESCRIPTION
**Pull request by kalidiagne**
# Description
This PR fixes problem with `autologin`, in detail
- Fix sign for a message that's shown immediately on opening the page
- If the user previously logged with `Unlock`, the login form will be shown if restart the flow 

# Issues

Fixes #8389 
Refs #

https://user-images.githubusercontent.com/20865711/159791097-2efd898f-e28b-4481-b318-bb0b8f13d6be.mov

# Checklist:

- [ ] 1 PR, 1 purpose: my Pull Request applies to a single purpose
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the docs to reflect my changes if applicable
- [ ] I have added tests (and stories for frontend components) that prove my fix is effective or that my feature works
- [ ] I have performed a self-review of my own code
- [ ] If my code involves visual changes, I am adding applicable screenshots to this thread

### CODE REVIEW

File: unlock-app/src/components/interface/checkout/Checkout.tsx [hunk 1]
```diff
@@ -147,11 +158,8 @@ export const Checkout = ({
   }
 
   const setCheckoutState = (state: string) => {
-    if (!state || state === 'connect' || state === 'loading') {
-      setShowBack(false)
-    } else {
-      setShowBack(true)
-    }
+    const showBack = !(!state || state === 'connect' || state === 'loading')
```
Comment (julien51) [inline unlock-app/src/components/interface/checkout/Checkout.tsx:161]:
```suggestion
    const showBack = state && state !== 'connect' && state !== 'loading'
```

File: packages/unlock-js/src/Unlock/v10/createLock.js [hunk 2]
```diff
@@ -35,7 +35,7 @@ export default async function (lock, callback) {
     maxNumberOfKeys = ETHERS_MAX_UINT
   }
   if (expirationDuration === -1) {
-    maxNumberOfKeys = ETHERS_MAX_UINT
+    expirationDuration = ETHERS_MAX_UINT
```
Comment (julien51) [inline packages/unlock-js/src/Unlock/v10/createLock.js:38]: Do you mind moving this to another PR please?

File: unlock-app/src/components/creator/CreatorLockForm.js [hunk 3]
```diff
@@ -87,12 +88,20 @@ const CreatorLockForm = ({ hideAction, lock, saveLock }) => {
     validateAndDispatch(target, [{ name, value: value * (60 * 60 * 24) }])
   }
 
-  const handleUnlimitedClick = () => {
+  const handleUnlimitedNumbersOfKeys = () => {
```
Comment (julien51) [inline unlock-app/src/components/creator/CreatorLockForm.js:91]: same as above :)

**Review by julien51 — COMMENTED** [review]
Let's move the changes on the unlimited supply to another PR please?

File: unlock-app/src/components/interface/checkout/CardConfirmationCheckout.tsx [hunk 4]
```diff
@@ -59,10 +59,16 @@ export const CardConfirmationCheckout = ({
     checkingRecipient,
   } = useAdvancedCheckout()
 
-  const totalPrice: number = Object.values(
-    lock.fiatPricing.usd as number
-  ).reduce((s: number, x: number): number => s + x, 0) as number
-  const fee = totalPrice - lock.fiatPricing.usd.keyPrice
+  let totalPrice: number = 0
```
Comment (julien51) [inline unlock-app/src/components/interface/checkout/CardConfirmationCheckout.tsx:62]: Please move to another PR (and/or rebase because I think I fixed this)

**Comment by kalidiagne** [comment]
PR updated. I also added a little video on the description

**Comment by julien51** [comment]
Thanks Kali!

Can you please test the following just to make sure:
- Checkout URL with no signature in it and user is using crypto wallet
- Checkout URL with no signature in it and user is using unlock account
- Checkout URL with signature required (add `messageToSign` to the `paywallConfig` object) and user is using crypto wallet)
- Checkout URL with signature required (add `messageToSign` to the `paywallConfig` object) and user is using unlock account

To test I think the best way is to connect these account on the dashboard (maybe keychain?).

File: packages/unlock-js/src/Unlock/v10/createLock.js [hunk 5]
```diff
@@ -35,7 +35,7 @@ export default async function (lock, callback) {
     maxNumberOfKeys = ETHERS_MAX_UINT
   }
   if (expirationDuration === -1) {
-    expirationDuration = ETHERS_MAX_UINT
+    maxNumberOfKeys = ETHERS_MAX_UINT
```
Comment (julien51) [inline packages/unlock-js/src/Unlock/v10/createLock.js:38]: Please let's move this into another PR :)

**Comment by julien51** [comment]
Oh, also can you check that it is persistent on the "oauth" flow? 
You can test with https://staging-app.unlock-protocol.com/checkout?client_id=ouvre-boite.com&redirect_uri=https://ouvre-boite.com/ or http://localhost:3000/checkout?client_id=ouvre-boite.com&redirect_uri=https://ouvre-boite.com/

**Review by julien51 — APPROVED** [review]
LGTM. Please do a few more cleanups and manual tests tho just to make sure we're not misisng anything!

**Comment by kalidiagne** [comment]
> Thanks Kali!
> 
> Can you please test the following just to make sure:
> 
> * Checkout URL with no signature in it and user is using crypto wallet
> * Checkout URL with no signature in it and user is using unlock account
> * Checkout URL with signature required (add `messageToSign` to the `paywallConfig` object) and user is using crypto wallet)
> * Checkout URL with signature required (add `messageToSign` to the `paywallConfig` object) and user is using unlock account
> 
> To test I think the best way is to connect these account on the dashboard (maybe keychain?).

Tested all this flows, on all of this is working well! 
please share with me if you find something wrong!

File: unlock-app/src/components/interface/checkout/Lock.tsx [hunk 6]
```diff
@@ -55,11 +55,7 @@ export const Lock = ({
 
   const alreadyHasKey = (key: any) => {
     const now = new Date().getTime() / 1000
-    if (key && (key.expiration === -1 || key.expiration > now)) {
-      setHasValidKey(true)
-    } else {
-      setHasValidKey(false)
-    }
+    setHasValidKey(key && (key.expiration === -1 || key.expiration > now))
```
Comment (searchableguy) [inline unlock-app/src/components/interface/checkout/Lock.tsx:58]: nit: Split this logic. It's not very readable.

File: unlock-app/src/pages/_app.tsx [hunk 7]
```diff
@@ -18,8 +18,8 @@ const UnlockApp = ({ Component, pageProps }: AppProps) => {
   }, [])
 
   return (
-    <GlobalWrapper pageProps={pageProps}>
```
Comment (searchableguy) [inline unlock-app/src/pages/_app.tsx:21]: I'm a bit worried about this. Have you checked if we pass any props to page aside from the checkout page and confirmed there is nothing dependent on it.

**Review by searchableguy — APPROVED** [review]
Looks fine.

File: unlock-app/src/components/creator/CreatorLockForm.js [hunk 8]
```diff
@@ -176,6 +176,7 @@ const CreatorLockForm = ({ hideAction, lock, saveLock }) => {
           <FormLockDuration>
             <input
               type="text"
+              inputMode="numeric"
```
Comment (julien51) [inline unlock-app/src/components/creator/CreatorLockForm.js:179]: Please move to another PR

File: unlock-app/src/components/interface/checkout/Checkout.tsx [hunk 9]
```diff
@@ -108,12 +115,24 @@ export const Checkout = ({
   const [existingKeys, setHasKey] = useReducer(keysReducer, {})
   const [selectedLock, selectLock] = useState<any>(null)
   const [savedMetadata, setSavedMetadata] = useState<any>(false)
+  const [storedLoginEmail, setStoredLoginEmail] = useState<string>('')
```
Comment (julien51) [inline unlock-app/src/components/interface/checkout/Checkout.tsx:118]: Maybe we could move that inside the `useAutoLogin` hook?

**Review by julien51 — APPROVED** [review]
Ok, I really want to test this on a draft deploy by netlify and I am excited to see it live!

File: unlock-app/src/pages/_app.tsx (hunk 7 above)
Comment (kalidiagne) [reply unlock-app/src/pages/_app.tsx:21]: is not used in other parts, I also checked the previous PR to see when this was introduced.

ref:
[7929](https://github.com/unlock-protocol/unlock/pull/7929)
[8010](https://github.com/unlock-protocol/unlock/pull/8010)

File: unlock-app/src/components/interface/checkout/Checkout.tsx (hunk 9 above)
Comment (kalidiagne) [reply unlock-app/src/components/interface/checkout/Checkout.tsx:112]: I defined the `storedLoginEmail` state in this component so we don't lose the value, we use the hook only to get the value

**Comment by julien51** [comment]
ok let's merge please!
